src/core/thread/mle.cpp

```diff
@@ -4018,6 +4027,16 @@ void Mle::HandleChildUpdateResponse(RxInfo &aRxInfo)
         {
         case kErrorNone:
             mTimeout = timeout;
+            if (mTimeout == 0 && mDetachGracefullyCallback != nullptr)
+            {
+                mDetachGracefullyTimer.Stop();
+                Stop();
+                SetTimeout(mTimeoutBeforeDetaching);
+                mDetachGracefullyCallback(mDetachGracefullyContext);
+                mDetachGracefullyCallback = nullptr;
+                mDetachGracefullyContext  = nullptr;
+                mTimeoutBeforeDetaching   = 0;
```

Comment:
better clear `mDetachGracefullyCallback` and  `mDetachGracefullyContext` first and then invoke the callback

Reply:
Done.